### ISSUE TITLE
"insecure string pickle" error preventing successful jobs reaching completed status when DiracFile used in output

### ISSUE THREAD

**Issue by twhyntie** [ganga-devs/ganga#1200]
Hello,

I'm running Ganga from the `development` branch (checked out yesterday) and trying to put my output from the job(s) onto a storage element with `DiracFile` (my output will ultimately be too large for using the sandbox).

The job runs and completes fine, and the output is written to the SE (I can check all of this from the DIRAC web portal). However, my Ganga instance can't seem to do the post-processing on the jobs and so reports them as failed, throwing the following error:

```
WARNING  An error occured finalising job: 63
WARNING  Attemting again (4 of 5) after 2.5-sec delay
INFO     job 63 status changed to "submitted"
INFO     job 63 status changed to "running"
INFO     job 63 status changed to "completing"
INFO     Contacting DIRAC for job: 63
WARNING  An error occured finalising job: 63
WARNING  Attemting again (5 of 5) after 2.5-sec delay
ERROR    Unable to finalise job after 63 retries due to error:
insecure string pickle
INFO     Forcing job 63 to status "failed"
INFO     Job 63 PostProcessor Failed
INFO     job 63 status changed to "failed"
ERROR    Exception raised executing 'job_finalisation' in Thread 'Job 63 Finalizing':
Traceback (most recent call last):
  File "/home/gridpp/sciencing27/lib/python2.7/site-packages/ganga-6.7.4-py2.7.egg/GangaCore/Core/GangaThread/WorkerThreads/WorkerThreadPool.py", line 115, in __worker_thread
    result = item.command_input.function(*these_args, **item.command_input.kwargs)
  File "/home/gridpp/sciencing27/lib/python2.7/site-packages/ganga-6.7.4-py2.7.egg/GangaDirac/Lib/Backends/DiracBase.py", line 1047, in job_finalisation
    DiracBase._internal_job_finalisation(job, updated_dirac_status)
  File "/home/gridpp/sciencing27/lib/python2.7/site-packages/ganga-6.7.4-py2.7.egg/GangaDirac/Lib/Backends/DiracBase.py", line 919, in _internal_job_finalisation
    job.backend.normCPUTime, getSandboxResult, file_info_dict, completeTimeResult = execute("finished_job(%d, '%s')" % (job.backend.id, output_path), cred_req=job.backend.credential_requirements)
  File "/home/gridpp/sciencing27/lib/python2.7/site-packages/ganga-6.7.4-py2.7.egg/GangaDirac/Lib/Utilities/DiracUtilities.py", line 245, in execute
    update_env=update_env)
  File "/home/gridpp/sciencing27/lib/python2.7/site-packages/ganga-6.7.4-py2.7.egg/GangaCore/Utility/execute.py", line 280, in execute
    stdout_temp = pickle.loads(stdout)
ValueError: insecure string pickle
```

Googling the `ValueError: insecure string pickle` error throws up a number of possible reasons, but might you have any thoughts on the problem? It only happens when a DiracFile is used in the `outputfiles` - if the output file is a `LocalFile` there's no error and Ganga reports the job as `completed`.

Thanks, @twhyntie

PS: I used the `development` branch after the problems with using the CVMFS version before - is this the right thing to do for now?

**Comment by rob-c** [comment]
@twhyntie I think the dev branch may be the best solution as I don't know if my DIRAC-UI fix went in or not to a release yet.

It looks like you're getting an incomplete pickle stream here at best guess.

First stab at helping. You're using Python 2.7 which seems a little brave. I don't know what the status of DIRAC/Ganga is with Python2.7 but have you tried using Python 2.6.


Failing this I would ask if you could edit `/home/gridpp/sciencing27/lib/python2.7/site-packages/ganga-6.7.4-py2.7.egg/GangaCore/Utility/execute.py` and add a line to print the content of stdout ahead of line 280.
There are 2 main possible problems here:
1) DIRAC is throwing an error which needs to be debugged
2) Ganga needs to add more protection around the complete flushing of buffers before unpickling them
3) Some combination of the above

What sort of environment are you trying to run this from a Centos7 or 6 derivative?

**Comment by twhyntie** [comment]
Thanks @rob-c 

I'm using a CernVM, which I believe is CentOS 6 derived. I'm using Python 2.7 because I couldn't get Ganga to build in 2.6 as there were some package dependencies that were no longer supported in 2.6 (I can't remember which ones at the moment). Plus there's the little message when Ganga starts kept telling me that I should move to 2.7 😉 

So I can't easily try with 2.6 at the moment.

But thanks for the `execute.py` tip - of course I can play with the output myself, so I'll see what's going on before line 280...

**Comment by twhyntie** [comment]
Update!

#GoodNews: printing `stdout` revealed the source of the error, which was an understood issue with the gfal tools (`dirac_ui/bashr` sets some bad environment variables, but it's easily fixable by setting up a grid UI properly - which I wasn't).

#BadNews (well, #IndifferentNews): I don't understand why the `stdout` message that told me this threw that particular error, so yeah I think there might be some more protection needed...

I'm happy for this to be closed for now (given I'm in 2.7 and it's now working for me) but perhaps worth flagging for future reference.

Thanks again for your help, @twhyntie

**Comment by rob-c** [comment]
@twhyntie  OK, great to hear this could be solved externally,

I think it's worth keeping this open and wrapping the sensitive code to return a more sensible error to help users in future. I'm going through doing some polish in my spare time and this is the sort of thing that could do with a sensible fix.
